In [1]:
import numpy as np
import pandas as pd
!pip install -q pyomo
from pyomo.environ import *
!apt-get install -y -qq coinor-cbc
opt_cbc=SolverFactory('cbc')

     |████████████████████████████████| 9.7 MB 20.4 MB/s 
     |████████████████████████████████| 49 kB 5.6 MB/s 
Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 159447 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to un

In [2]:
data=np.loadtxt('lab7_ex2_lp_ip_coef.txt',delimiter=',')

In [3]:
model=ConcreteModel()
m=len(data[0,:])-1
n=len(data[:,0])-1

In [4]:
row_indices=np.arange(m)
col_indices=np.arange(n)

In [5]:
model.x=Var(row_indices,domain=Binary)

In [6]:
model.objective=Objective(expr=sum(model.x[i]*data[0][i] for i in row_indices),sense=maximize)

In [7]:
model.constraints=ConstraintList()

In [8]:
for j in col_indices:
  model.constraints.add(expr=sum(data[j+1,i]*model.x[i] for i in row_indices) <=data[j+1,-1])

In [9]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :  500 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 15

In [10]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
  # display solution
print('\nObjective = ', model.objective())
result.solver.time
print(result)

Solver status: ok
Solver termination condition: optimal

Objective =  295828.0

Problem: 
- Name: unknown
  Lower bound: 295828.0
  Upper bound: 295828.0
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 500
  Number of binary variables: 500
  Number of integer variables: 500
  Number of nonzeros: 500
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 50.02
  Wallclock time: 53.84
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 347044
      Number of created subproblems: 347044
    Black box: 
      Number of iterations: 1407089
  Error rc: 0
  Time: 53.88973903656006
Solution: 
- number of solutions: 0
  number of solutions displayed: 0



In [11]:
print('solver time=',result.solver.time)

53.88973903656006


$\huge{solution \ \  of \ \  2.3}$

objective=295828.0

$\huge{solution \ \ of \ \ 2.4}$

solver time = 79.31881546974182

time is different when we run it again

Branch and bound: 
      
Number of bounded subproblems: 347044
    
Number of created subproblems: 347044

In [12]:
model.x.domain=Reals

In [13]:
model.x.setlb(0)
model.x.setub(1)

In [14]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
  # display solution
print('\nObjective = ', model.objective())
result.solver.time
print(result)

Solver status: ok
Solver termination condition: optimal

Objective =  295896.37697236

Problem: 
- Name: unknown
  Lower bound: 295896.377
  Upper bound: 295896.377
  Number of objectives: 1
  Number of constraints: 6
  Number of variables: 501
  Number of nonzeros: 500
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.01
  Wallclock time: 0.01
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 18
  Error rc: 0
  Time: 0.01818060874938965
Solution: 
- number of solutions: 0
  number of solutions displayed: 0



In [15]:
print('solver time=',result.solver.time)

0.01818060874938965


$ \huge{solution \ \ of \ \   2.5}$

Objective =  295896.37697236

$ \huge{solution \ \ of \ \ \ 2.6}$

solver time=0.054627180099487305
time is different when we run it again

in MILP it takes more time compare to LP model
since in MILP it is solved by branch and bound algorithm so we have to check each nodes for solution thats not happend in LP problem if any problem has feasible solution in LP then it is fastest way of solving MILP

A branch and bound algorithm consist of stepwise enumeration of possible candidate solutions by exploring the entire search space. With all the possible solutions, we first build a rooted decision tree. The root node represents the entire search space 

clearly for MILP system has created 347044 subproblem and bounded 347044 subproblem but in LP it is not happend so it is faster than MILP


$\huge{solution \ \ of \ \  2.7}$


  Statistics: 

  for lp 
Branch and bound:


Number of bounded subproblems: None

Number of created subproblems: None

for milp 

Branch and bound: 
      
Number of bounded subproblems: 347044
    
Number of created subproblems: 347044